[Merged by Bors] - feat(ring_theory/polynomial/chebyshev): chebyshev polynomials of the first kind

jcommelin opened the pull request:
If T_n denotes the n-th Chebyshev polynomial of the first kind, then the
polynomials 2*T_n(X/2) form a Lambda structure on Z[X].

I call these polynomials the lambdashev polynomials, because, as far as I
am aware they don't have a name in the literature.

We show that they commute, and that the p-th polynomial is congruent to X^p
mod p. In other words: a Lambda structure.



---

<!-- put comments you want to keep out of the PR commit here -->

- [x] depends on: #4241
- [x] depends on: #4578 
- [x] depends on: #4595
- [x] depends on: #4617
- [x] depends on: #4656

jjaassoonn committed:
feat(polynomial*) : add a lemma stating that
a poylnomial is zero if it is zero on an interval

jjaassoonn committed:
add doc_string

jjaassoonn committed:
generalize a bit

jjaassoonn committed:
delete an unnecessary variable

jjaassoonn committed:
shorten proof a little bit

rwbarton committed:
feat(data/set/intervals/infinite): intervals are infinite

jcommelin committed:
feat(ring_theory/polynomial/chebyshev): chebyshev polynomials of the first kind

If T_n denotes the n-th Chebyshev polynomial of the first kind,
then the polynomials 2*T_n(X/2) form a Lambda structure on Z[X].

I call these polynomials the lambdashev polynomials,
because, as far as I am aware they don't have a name in the literature.

We show that they commute, and that the p-th polynomial is congruent to
X^p mod p. In other words: a Lambda structure.

rwbarton committed:
refactor proofs

jcommelin committed:
Merge branch 'master' into chebpoly

jcommelin committed:
Merge remote-tracking branch 'origin/intervals-infinite' into chebpoly

jcommelin committed:
clean up

jcommelin committed:
Merge branch 'intervals-infinite' into chebpoly

Commit:
fix build

robertylewis committed:
fix again

jcommelin committed:
Merge branch 'master' into chebpoly

jjaassoonn committed:
add a lemma `polynomial_zero_of_zero_on_infinite_set`

jjaassoonn committed:
resolve confilict

jjaassoonn committed:
remove unnecessary file

jjaassoonn committed:
change name

jcommelin committed:
fix? no fix :sad:

jjaassoonn committed:
apply suggestion from @jcommelin

jcommelin committed:
Revert "fix? no fix :sad:"

This reverts commit 88853a4c4d3cec511b8415b6b7274cb663e38b43.

jcommelin committed:
Revert "fix again"

This reverts commit 8c70eb71c2da6f57ce46d35b442126a99855b875.

jcommelin committed:
Revert "fix build"

This reverts commit 8cfde48d51463b7beb920526467a20531200ce02.

jcommelin committed:
Merge remote-tracking branch 'origin/zero_on_interval_imp_zero_polynomial' into chebpoly

hrmacbeth commented on src/ring_theory/polynomial/chebyshev.lean:
```diff
@@ -0,0 +1,301 @@
+section complex
+open complex
+
+/-- the `n`-th Chebyshev polynomial evaluates on `cos θ` to the value `n * cos θ` -/
```
```suggestion
/-- the `n`-th Chebyshev polynomial evaluates on `cos θ` to the value `cos (n * θ)` -/
```

hrmacbeth commented on src/ring_theory/polynomial/chebyshev.lean:
```diff
@@ -0,0 +1,301 @@
+
+/-- the `n`-th Chebyshev polynomial evaluates on `cos θ` to the value `n * cos θ` -/
+@[simp] lemma chebyshev₁_complex_cos (θ : ℂ) :
+  ∀ n, (chebyshev₁ ℂ n).eval (cos θ) = cos (n * θ)
```
One could argue that the `simp` should go in the other direction!  (I think @PatrickMassot was thinking along these lines.). That is, that the relevant simp-ing of Chebyshev polynomials is to let you reduce the number of different arguments fed to `cos` within a single expression.  Is there a reason you chose this direction for your simp?

hrmacbeth replied:
Also, perhaps we could move this `complex` section to `analysis.special_functions.trigonometric`?  It'd be nice to be able to refer to the Chebyshev polynomials there.  (I see that a later generic lemma `chebyshev₁_mul ` depends on the complex section, but is there another way to do it?)

jcommelin replied:
I've done this.

jcommelin committed:
Merge branch 'master' into chebpoly

jcommelin committed:
feat(algebra/char_zero): char_zero.infinite

jcommelin committed:
feat(algebra/char_p): fields with a hom between them have same char

jcommelin committed:
feat(analysis/special_functions/trigonometric): range_cos_infinite

jcommelin committed:
feat(data/polynomial/eval): easy lemmas + speedup

jcommelin committed:
Merge branches 'chebpoly-preps', 'chebpoly-preps-2', 'chebpoly-preps-3' and 'chebpoly-preps-4' into chebpoly

jcommelin committed:
undo import change

jcommelin committed:
change direction of simp-lemma

jcommelin committed:
Update src/analysis/special_functions/trigonometric.lean

Co-authored-by: Anatole Dedecker <anatolededecker@gmail.com>

jcommelin committed:
factor out sublemmas; copy for sin

jcommelin committed:
Merge branch 'master' into chebpoly

jcommelin committed:
reorganize

jcommelin committed:
fix long lines

jcommelin committed:
fix build

jcommelin committed:
fix build

jcommelin committed:
wip

jcommelin committed:
refactor(algebra/quadratic_discriminant): drop linearity condition; cleanup

Renames:

- `discriminant_le_zero` to `discrim_le_zero`
- `discriminant_lt_zero` to `discrim_lt_zero`

jcommelin committed:
weaken from lin.ord.fld to lin.ord.semiring

jcommelin committed:
Merge branch 'quad-disc' into chebpoly-preps-3

jcommelin committed:
this should have been gone already

jcommelin committed:
Update quadratic_discriminant.lean

jcommelin committed:
finish proofs

jcommelin committed:
Merge branch 'quad-disc' into chebpoly-preps-3

jcommelin committed:
generalise

jcommelin committed:
Merge branch 'chebpoly-preps-3' into chebpoly

jcommelin committed:
fix build

jcommelin committed:
Merge branch 'master' into chebpoly

kim-em commented on src/analysis/special_functions/trigonometric.lean:
```diff
@@ -1926,6 +1948,32 @@ sin_surjective.range_eq
+/-- `cos (n * θ)` is equal to the `n`-th Chebyshev polynomial evaluated on `cos θ`. -/
+@[simp] lemma cos_nat_mul (n : ℕ) (θ : ℂ) :
+  cos (n * θ) = (chebyshev₁ ℂ n).eval (cos θ) :=
+(chebyshev₁_complex_cos θ n).symm
```
I'm dubious about this being a `simp` lemma.

jcommelin replied:
Fair enough. I removed the attribute.

jcommelin committed:
fix build, remove simp attribute

kim-em commented on src/ring_theory/polynomial/chebyshev/basic.lean:
```diff
@@ -0,0 +1,204 @@
+lemma lambdashev_zmod_p (p : ℕ) [fact p.prime] :
+  lambdashev (zmod p) p = X ^ p :=
+begin
+  -- Recall that `lambdashev_eval_add_inv` characterise `lambdashev R p`
```
```suggestion
  -- Recall that `lambdashev_eval_add_inv` characterises `lambdashev R p`
```

kim-em commented on src/ring_theory/polynomial/chebyshev/defs.lean:
```diff
@@ -0,0 +1,115 @@
+  chebyshev₁ R (n + 2) = 2 * X * chebyshev₁ R (n + 1) - chebyshev₁ R n :=
+by rw chebyshev₁
+
+lemma chebyshev₁_two_le (n : ℕ) (h : 2 ≤ n) :
```
```suggestion
lemma chebyshev₁_of_two_le (n : ℕ) (h : 2 ≤ n) :
```

kim-em replied:
Or `chebyshev₁_eq_of_two_le`?

kim-em commented on src/ring_theory/polynomial/chebyshev/basic.lean:
```diff
@@ -0,0 +1,204 @@
+        exact ⟨_, 1, rfl, one_ne_zero⟩ },
+      { simp only [hx, or_false, exists_eq_right],
+        exact ⟨_, rfl, hx⟩ } } }
+end
```
It's so cool that you can actually carry out this argument without a too-unpleasant de Bruijn factor!

jcommelin replied:
Yes, I was very pleased by this myself.

jcommelin commented on src/ring_theory/polynomial/chebyshev/defs.lean:
```diff
@@ -0,0 +1,115 @@
+  lambdashev R (n + 2) = X * lambdashev R (n + 1) - lambdashev R n :=
+by rw lambdashev
+
+lemma lambdashev_two_le (n : ℕ) (h : 2 ≤ n) :
```
```suggestion
lemma lambdashev_eq_two_le (n : ℕ) (h : 2 ≤ n) :
```

jcommelin committed:
Apply suggestions from code review

Co-authored-by: Scott Morrison <scott.morrison@gmail.com>
